internal/cmd/beta/security-group-rule/create/create_test.go

```diff
@@ -0,0 +1,353 @@
+		mod(&request)
+	}
+	return request
+}
+
+func fixtureRequiredRequest(mods ...func(request *iaas.ApiCreateSecurityGroupRuleRequest)) iaas.ApiCreateSecurityGroupRuleRequest {
+	request := testClient.CreateSecurityGroupRule(testCtx, testProjectId, testSecurityGroupId)
+	request = request.CreateSecurityGroupRulePayload(iaas.CreateSecurityGroupRulePayload{
+		Direction: utils.Ptr("ingress"),
+	})
+	for _, mod := range mods {
+		mod(&request)
+	}
+	return request
+}
+
+func fixturePayload(mods ...func(payload *iaas.CreateSecurityGroupRulePayload)) iaas.CreateSecurityGroupRulePayload {
+	payload := iaas.CreateSecurityGroupRulePayload{
+		Direction:   utils.Ptr("ingress"),
+		Description: utils.Ptr("example-description"),
+		Ethertype:   utils.Ptr("ether"),
+		IcmpParameters: &iaas.ICMPParameters{
+			Code: utils.Ptr(int64(0)),
+			Type: utils.Ptr(int64(8)),
+		},
+		IpRange: utils.Ptr("10.1.2.3"),
+		PortRange: &iaas.PortRange{
+			Max: utils.Ptr(int64(24)),
+			Min: utils.Ptr(int64(22)),
+		},
+		Protocol: &iaas.CreateProtocol{
+			Int64:  utils.Ptr(int64(1)),
+			String: utils.Ptr("icmp"),
+		},
+		RemoteSecurityGroupId: utils.Ptr(testRemoteSecurityGroupId),
+	}
+	for _, mod := range mods {
+		mod(&payload)
+	}
+	return payload
+}
+
+func TestParseInput(t *testing.T) {
+	tests := []struct {
+		description   string
+		flagValues    map[string]string
+		isValid       bool
+		expectedModel *inputModel
+	}{
+		{
+			description: "base",
+			flagValues: fixtureFlagValues(func(flagValues map[string]string) {
+				delete(flagValues, portRangeMaxFlag)
+				delete(flagValues, portRangeMinFlag)
+				delete(flagValues, protocolNumberFlag)
+			}),
+			isValid: true,
+			expectedModel: fixtureInputModel(func(model *inputModel) {
+				model.PortRangeMax = nil
+				model.PortRangeMin = nil
+				model.ProtocolNumber = nil
+			}),
+		},
+		{
+			description: "required only",
+			flagValues: fixtureFlagValues(func(flagValues map[string]string) {
+				delete(flagValues, descriptionFlag)
+				delete(flagValues, etherTypeFlag)
+				delete(flagValues, icmpParameterCodeFlag)
+				delete(flagValues, icmpParameterTypeFlag)
+				delete(flagValues, ipRangeFlag)
+				delete(flagValues, portRangeMaxFlag)
+				delete(flagValues, portRangeMinFlag)
+				delete(flagValues, remoteSecurityGroupIdFlag)
+				delete(flagValues, protocolNumberFlag)
+				delete(flagValues, protocolNameFlag)
+			}),
+			isValid: true,
+			expectedModel: fixtureInputModel(func(model *inputModel) {
+				model.Description = nil
+				model.EtherType = nil
+				model.IcmpParameterCode = nil
+				model.IcmpParameterType = nil
+				model.IpRange = nil
+				model.PortRangeMax = nil
+				model.PortRangeMin = nil
+				model.RemoteSecurityGroupId = nil
+				model.ProtocolNumber = nil
+				model.ProtocolName = nil
+			}),
+		},
+		{
+			description: "direction missing",
+			flagValues: fixtureFlagValues(func(flagValues map[string]string) {
+				delete(flagValues, directionFlag)
+			}),
+			isValid: false,
```

Comment:
If set to true, this test will fail because of `error validating flag groups: if any flags in the group [protocol-number protocol-name] are set none of the others can be; [protocol-name protocol-number]` but it should fail because of "direction missing" (false positive).

This is a general issue with the tests. I would like to have tests which expect e.g. that "direction is missing" and if something has been forgotten the test should fail. I will raise this issue in the team.

Reply:
adapted unit test that, if isValid is set to true, it will fail with the correct error message